toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,91 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(bool first_name_component,
+                                              llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
+    if (parent.inst_id == SemIR::InstId::PackageNamespace) {
+      break;
+    }
+    if (!first_name_component) {
+      os << '.';
+    }
+    first_name_component = false;
+    CARBON_KIND_SWITCH(sem_ir.insts().Get(parent.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        const auto& impl = sem_ir.impls().Get(impl_decl.impl_id);
+        if (auto opt_class_self =
+                types.TryGetAs<SemIR::ClassType>(impl.self_id)) {
+          MangleInverseQualifiedNameScope(
+              true, os,
+              sem_ir.classes().Get(opt_class_self->class_id).scope_id);
+        } else {
+          auto builtin_self = types.GetAs<SemIR::BuiltinInst>(impl.self_id);
+          os << builtin_self.builtin_inst_kind.label();
+        }
+        os << ':';
+        auto opt_interface_constraint =
```

Comment:
What's the "opt_" mean here?

Reply:
Optional - not sure if there's another convention here. I'd prefer something at the declaration that indicates the optionality (either using an explicit type instead of `auto` or something in the name like `opt_`)

Reply:
I think `GetAs` doesn't return an optional.

Note, typically we just don't remark on it. But I think you should take a closer look here.

Reply:
ah, yeah, must've been some lingering thing. Fixed/removed the prefix.